### PR TITLE
(#3170) sitesection must have unique url.

### LINKED ISSUE #3170: Site section under a parent site section must have an unique pretty url
### Issue description
*Description of the issue*
When preceded by the parent section's pretty URL path, site section should not be allowed to have duplicated Purls
> **ESTIMATE** TBD

### Steps to reproduce the issue

1.  Go to Structure -> taxonomy -> site sections in cms
2. select home's children
2.  modify one of the children to have same Purl as the other chid under same parent 
3.  save and - system allows it 

### What's the expected result?
- the system should error that only unique Purls are allowed and not let the section to be saved
- The error message should be, "The section is unable to be saved. Please enter a unique pretty url."

### What's the actual result?
- it saves with no problem which results in duplicate site section urls

### Additional details / screenshot
- Screenshot
Screen Shot 2021-09-16 at 5 24 59 PM

### Related Tickets
* Original ticket with requirements where this bug was missed https://github.com/NCIOCPL/cgov-digital-platform/issues/175
*  PR that had a comment mentioning that this issue was noted and will need a bug ticket to fix it later https://github.com/NCIOCPL/cgov-digital-platform/pull/589

### PR DESCRIPTION
**Pull request by navsunka**
Closes #3170

### CODE REVIEW

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * )
+ */
+class ValidatePrettyUrlConstraint extends Constraint {
+
+  /**
+   * Unique pretty url.
+   *
+   * @var string
+   */
+  public $uniquePrettyUrl = 'The site section cannot be saved. Site sections must have unique pretty url.';
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php:22]: @andyvanavery31 and @sarinapadilla - would one of you please chime in with text for the error message?  [The original ticket](https://github.com/NCIOCPL/cgov-digital-platform/issues/175) only specifies that they must be unique.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      // Exclude the current term if it exists.
+      if (!$entity->isNew()) {
+        $query->condition('tid', $entity->id(), '!=');
+      }
+      $result = $query->execute();
+      if (!empty($result)) {
+        $this->context->addViolation($constraint->uniquePrettyUrl);
+      }
+    }
+  }
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php:70]: Should this be returning something?  Line 51 above returns `NULL` if `$constraint` isn't the right type.

**Review by blairlearn — CHANGES_REQUESTED** [review]

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * A composite constraint validating Site Section pretty url.
+ *
+ * @Constraint(
+ *   id = "CgovValidatePrettyUrlConstraint",
+ *   label = @Translation("ValidatePrettyUrlConstraint", context="Validation"),
+ * )
+ */
+class ValidatePrettyUrlConstraint extends Constraint {
```
[the remark this answers is not in the document]
Comment (navsunka) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php]: I have changed the name to UniquePrettyUrl and this UniqueCgovUrlConstraint is checking for node type not for taxonomy.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Validates the ValidatePrettyUrlConstraintValidator constraint.
+ */
+class ValidatePrettyUrlConstraintValidator extends ConstraintValidator implements ContainerInjectionInterface {
```
[the remark this answers is not in the document]
Comment (navsunka) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php]: I have changed the name to UniquePrettyUrlValidator and this UniqueCgovUrlConstraintValidator is checking for node type not for taxonomy.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  public function validate($field, Constraint $constraint) {
+
+    if (!($constraint instanceof ValidatePrettyUrlConstraint)) {
+      return NULL;
+    }
+
+    $entity = $field->getEntity();
+    $prettyUrl = $entity->field_pretty_url->value;
+    if (!empty($prettyUrl)) {
```
[the remark this answers is not in the document]
Comment (navsunka) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php]: Fixed.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php (hunk 2 above)
Comment (navsunka) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php]: It should not written anything, I have written same as in drupal core constraint validators.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php (hunk 1 above)
Comment (andyvanavery31) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraint.php]: @navsunka @blairlearn updated the ticket #3170 with text for the error message in the Expected Results section.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php (hunk 2 above)
Comment (blairlearn) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidatePrettyUrlConstraintValidator.php]: You are correct, it should not be returning anything.  Looking at how the `validate` method is declared in `Symfony\Component\Validator\ConstraintValidatorInterface`
([link](https://github.com/symfony/symfony/blob/v6.3.12/src/Symfony/Component/Validator/ConstraintValidatorInterface.php#L32)), the method is void and arguably it's the return NULL which should really just be `return`.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/cgov_site_section.module [hunk 6]
```diff
@@ -776,9 +776,13 @@ function cgov_site_section_entity_bundle_field_info_alter(&$fields, EntityTypeIn
   }
   if ($entity_type->id() == 'taxonomy_term'  && $bundle === 'cgov_site_sections') {
     if (isset($fields['parent'])) {
-      // Add custom constraint to this field
+      // Add custom constraint to this field.
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/cgov_site_section.module:779]: Please replace this comment with something more descriptive. A good comment should explain the _intent_ of the code.  In this case, what rule is being validated? That a parent exists? That there's only one?

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/cgov_site_section.module [hunk 7]
```diff
@@ -776,9 +776,13 @@ function cgov_site_section_entity_bundle_field_info_alter(&$fields, EntityTypeIn
...
       $fields['parent']->addConstraint('CgovValidateParentConstraint', []);
     }
+    if (isset($fields['field_pretty_url'])) {
+      // Add custom constraint to pretty url.
+      $fields['field_pretty_url']->addConstraint('CgovValidatePrettyUrlConstraint', []);
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/cgov_site_section.module:784]: Same thing as above.

**Review by blairlearn — CHANGES_REQUESTED** [review]
The new constraint classes duplicate the functionality of the existing `UniqueCgovUrlConstraint` and `UniqueCgovUrlConstraintValidator` classes.

Technically, yes, the existing classes validate for `node` instead of `taxonomy_term`, however they are enforcing the same rule, just for different entity types.  This should be done with the existing classes.

The new classes also introduce a bug where only a single site section can have a given pretty url value, regardless of where it exists in the hierarchy.  (e.g. You cannot create a site section with a pretty url value of "funding" under "news' if another site section with the value of "funding" already exists under "research".)

The validation logic should be along the lines of

1. Fetch the parent `cgov_site_section`
2. Get the parent's `computed_path`
3. Append the proposed `field_pretty_url` value
4. Check for the existence of any other `cgov_site_section` with that same `computed_path`.

You'll also need to make certain the new `computed_path` doesn't collide with the `computed_path` for anything else.

A specific scenario to check:

1. I make `/foo` site section
2. I make `/foo/bar` node, where the node has `bar` as its `field_pretty_url`
3. I then go make `/foo/bar` site section where `bar` is `field_pretty_url`

**Comment by KateMashkinaNIH** [comment]
this scenario does not pass - it allowed me to create a site section 
Scenario: System does not allow the user to create a site section where the site section pretty URL matches another node item's site section path + pretty URL
  Given the Research site section with path /research exists 
    And a node item within Research site section with pretty URL set to "areas" exists
    And the user is creating a new site section within Research site section
  When the user sets the site section pretty URL to "areas" and tries to save
  Then the system will not allow the user to save the site section

**Comment by KateMashkinaNIH** [comment]
one more thing - the site section name should be case insensitive, e.g. About Cancer and About cancer should be treated the same and system should not allow those two section to be created with the same purl

**Review by blairlearn — CHANGES_REQUESTED** [review]

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidateParentConstraintValidator.php [hunk 8]
```diff
@@ -135,14 +135,14 @@ public function getLanguageRoots() {
    *   Boolean.
    */
   public function checkChildUniqueName($formData, $parentItems) {
-    $name = $formData['name'][0]['value'];
+    $name = strtolower($formData['name'][0]['value']);
```
[the remark this answers is not in the document]
Comment (blairlearn) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_site_section/src/Plugin/Validation/Constraint/ValidateParentConstraintValidator.php:138]: From [Slack conversation](https://ncisape.slack.com/archives/D051JFA5LB1/p1729013864347149?thread_ts=1729010782.098829&cid=D051JFA5LB1), this change is related to the requirement that pretty url uniqueness should be case-insensitive (e.g. `chicken` and `Chicken` are treated as the same pretty  url).

**Review by blairlearn — APPROVED** [review]

**Pull request merged**